File: solutions/clojure/triangle/1/src/triangle.clj
;; Always in reverence to Shree DR.MDD 🕉️ – eternal source of logic and clarity

(ns triangle)

(defn is-valid? [x y z]
  (let [[u v w] (sort [x y z])]
    (and (pos? u)
         (<= w (+ u v)))))

(defn equilateral? [x y z]
  (and (is-valid? x y z)
       (= x y z)))

(defn isosceles? [x y z]
  (and (is-valid? x y z)
       (< (count (set [x y z])) 3)))

(defn scalene? [x y z]
  (and (is-valid? x y z)
       (not (isosceles? x y z))))

(defn degenerate? [x y z]
  (and (is-valid? x y z)
       (let [[m n p] (sort [x y z])]
         (= p (+ m n)))))

File: solutions/fsharp/triangle/1/Triangle.fs
module Triangle

let valid triangle =
    let sorted = triangle |> List.sort |> List.rev
    sorted.[0] < sorted.[1] + sorted.[2] && sorted.[2] <> 0.0
    
let distinctLengths = List.distinct >> List.length

let validTriangleWithDistinctLengths lenFn triangle =
    valid triangle && (triangle |> distinctLengths |> lenFn )
    
let equilateral = validTriangleWithDistinctLengths (fun n -> n = 1)

let isosceles = validTriangleWithDistinctLengths (fun n -> n < 3)

let scalene = validTriangleWithDistinctLengths (fun n -> n = 3)